README.md
# Social Auth Buttons Kit
Flutter UI Components for adding social sign in buttons in your project following the brand guidlines of each social provider.
Social Sign in is a part of almost every app and this makes it easier to add buttons quickly.

![Screenshot](https://raw.githubusercontent.com/botraunak/social_auth_btn_kit/v0.1.0/screenshots/demo.png)

## Supported Provider
- Facebook
- Google
- Apple


## Getting started

To get stated add this as a dependency in your project.
```
flutter pub add social_auth_btn_kit
```

## Usage

### Facebook
Facebook brand guides allow for a normal and outlined button
can be passed as the ```variant``` parameter ```FacebookTypeVariants```
```dart
enum FacebookTypeVariants {
  normal,
  outlined,
}
```
Also the theme variant can be configured using the ```theme``` parameter
```dart
enum FacebookThemeVariants {
  light,
  dark,
}
```

### Google
Google brand guides have three themed buttons can be passed as the ```theme``` parameter ```GoogleThemeVariants```
```dart
enum GoogleThemeVariants {
  light,
  dark,
  neutral,
}
```
### Apple
Apple button support is also added with outline and dark support according to brand


```dart
SocialAuthBtn.facebook(
                variant: FacebookTypeVariants.outlined,
                onPressed: () {
                  debugPrint("DEBUG: Facebook Btn Pressed");
                },
              ),
```

## Custom buttons
You can create your own buttons using these quick styles, just add the icon in your assets folder and call the main constructor with custom values
```dart
SocialAuthBtn(
                onPressed: () {
                  debugPrint("DEBUG: Instagram Btn Pressed");
                },
                icon: 'assets/instagram.png',
                text: 'Connect Instagram',
                borderSide: const BorderSide(),
              ),
```



CHANGELOG.md
## 0.1.0

* Added Screenshot 

## 0.0.1

* Added Support for Facebook, Google, Apple
* Added Custom Buttons Support with your own icon and styles


test/social_auth_btn_kit_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:social_auth_btn_kit/social_auth_btn_kit.dart';

void main() {
  test('adds one to input values', () {});
}


example/lib/demo.dart
import 'package:flutter/material.dart';
import 'package:social_auth_btn_kit/social_auth_btn_kit.dart';
import 'package:social_auth_btn_kit/social_auth_btn_variants.dart';

class DemoPage extends StatelessWidget {
  const DemoPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Social Auth Btn Kit Demo'),
      ),
      backgroundColor: Colors.white,
      body: Padding(
        padding: const EdgeInsets.all(16.0),
        child: ListView(
          children: [
            Center(
              child: SocialAuthBtn.facebook(
                variant: FacebookTypeVariants.outlined,
                onPressed: () {
                  debugPrint("DEBUG: Facebook Btn Pressed");
                },
              ),
            ),
            const SizedBox(height: 16),
            Center(
              child: SocialAuthBtn.facebook(
                onPressed: () {
                  debugPrint("DEBUG: Facebook Btn Pressed");
                },
              ),
            ),
            const SizedBox(height: 16),
            Center(
              child: SocialAuthBtn.facebook(
                theme: FacebookThemeVariants.dark,
                onPressed: () {
                  debugPrint("DEBUG: Facebook Btn Pressed");
                },
              ),
            ),
            const SizedBox(height: 16),
            Center(
              child: SocialAuthBtn.facebook(
                variant: FacebookTypeVariants.outlined,
                theme: FacebookThemeVariants.dark,
                onPressed: () {
                  debugPrint("DEBUG: Facebook Btn Pressed");
                },
              ),
            ),
            const SizedBox(height: 16),
            Center(
              child: SocialAuthBtn.google(
                onPressed: () {
                  debugPrint("DEBUG: Google Btn Pressed");
                },
              ),
            ),
            const SizedBox(height: 16),
            Center(
              child: SocialAuthBtn.google(
                onPressed: () {
                  debugPrint("DEBUG: Google Btn Pressed");
                },
                theme: GoogleThemeVariants.dark,
              ),
            ),
            const SizedBox(height: 16),
            Center(
              child: SocialAuthBtn.google(
                onPressed: () {
                  debugPrint("DEBUG: Google Btn Pressed");
                },
                theme: GoogleThemeVariants.neutral,
              ),
            ),
            const SizedBox(height: 16),
            Center(
              child: SocialAuthBtn.apple(
                onPressed: () {
                  debugPrint("DEBUG: Apple Btn Pressed");
                },
              ),
            ),
            const SizedBox(height: 16),
            Center(
              child: SocialAuthBtn.apple(
                onPressed: () {
                  debugPrint("DEBUG: Apple Btn Pressed");
                },
                variant: AppleTypeVariants.outlined,
              ),
            ),
            const SizedBox(height: 16),
            Center(
              child: SocialAuthBtn.apple(
                onPressed: () {
                  debugPrint("DEBUG: Apple Btn Pressed");
                },
                theme: AppleThemeVariants.black,
              ),
            ),
            const SizedBox(height: 16),
            Center(
              child: SocialAuthBtn(
                onPressed: () {
                  debugPrint("DEBUG: Instagram Btn Pressed");
                },
                icon: 'assets/instagram.png',
                text: 'Connect Instagram',
                borderSide: const BorderSide(),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/social_auth_method.dart
// Class that manages social auth methods
import 'package:flutter/material.dart';
import 'package:social_auth_btn_kit/social_auth_colors.dart';

enum ButtonTypes { google, facebook, apple }

enum ButtonModes { light, dark, neutral }

class SocialAuthMethod {
  const SocialAuthMethod({
    required this.text,
    required this.icon,
    this.backgroundColor = Colors.white,
    this.textColor = Colors.black,
  });

  factory SocialAuthMethod.get({
    required ButtonTypes type,
    ButtonModes mode = ButtonModes.light,
  }) {
    if (type == ButtonTypes.google) {
      const icon = 'assets/logos/google.png';
      final googleColors = GoogleColors.fromMode(mode: mode);
      return SocialAuthMethod(
        text: 'Sign in with Google',
        icon: icon,
        textColor: googleColors.lblColor,
        backgroundColor: googleColors.bgColor,
      );
    }
    if (type == ButtonTypes.facebook) {
      const icon = 'assets/logos/facebook.png';
      return const SocialAuthMethod(
        text: 'Login with Facebook',
        icon: icon,
        textColor: Colors.white,
        backgroundColor: facebookBgColorLight,
      );
    }
    if (type == ButtonTypes.apple) {
      const icon = 'assets/logos/apple.png';
      return const SocialAuthMethod(
        text: 'Sign in with Apple',
        icon: icon,
        textColor: Color(0xFF0E0E0E),
      );
    }
    throw Exception('Invalid Social Auth Method');
  }
  final String text;
  final String icon;
  final Color backgroundColor;
  final Color textColor;
}


lib/social_auth_btn_kit.dart
import 'package:flutter/material.dart';
import 'package:social_auth_btn_kit/social_auth_btn_variants.dart';
import 'package:social_auth_btn_kit/social_auth_colors.dart';

/// SocialAuthBtn
/// Base Btn which handles all social logins
class SocialAuthBtn extends StatelessWidget {
  /// Creates a new instance of `SocialAuthBtn`.
  /// The [onPressed] parameter must not be null.
  /// This is the callback that will be called when the button is pressed.
  /// The [text] parameter must not be null.
  /// This is the text that will be displayed on the button.
  /// The [icon] parameter must not be null.
  /// This is the icon that will be displayed on the button.
  /// The [iconColor] parameter defaults to `null`.
  /// This is the color of the icon.
  /// The [textColor] parameter defaults to `Colors.black`.
  /// This is the color of the text.
  /// The [backgroundColor] parameter defaults to `Colors.white`.
  /// This is the background color of the button.
  /// The [width] parameter defaults to 300.
  /// This is the width of the button.
  /// The [borderRadius] parameter defaults to 8.
  /// This is the border radius of the button.
  /// The [borderSide] parameter defaults to `BorderSide.none`.
  /// This is the border side of the button.
  /// The [loading] parameter defaults to `false`.
  /// This determines whether to show a loading indicator on the button.
  /// The [loaderColor] parameter defaults to `null`.
  /// This is the color of the loading indicator.
  const SocialAuthBtn({
    required this.text,
    required this.onPressed,
    required this.icon,
    this.iconColor,
    this.package,
    this.textColor = Colors.black,
    this.backgroundColor = Colors.white,
    this.width = 300,
    this.borderRadius = 8,
    this.borderSide = BorderSide.none,
    this.loading = false,
    this.loaderColor,
    super.key,
  });

  /// Creates a new instance of `SocialAuthBtn` for Facebook.
  ///
  /// The [onPressed] parameter must not be null.
  /// This is the callback that will be called when the button is pressed.
  ///
  /// The [variant] parameter defaults to `FacebookTypeVariants.normal`.
  /// This determines the type of Facebook button to display.
  ///
  /// The [theme] parameter defaults to `FacebookThemeVariants.light`.
  /// This determines the theme of the Facebook button.
  ///
  /// The button's border, background color, text color, and icon color
  /// are determined based on the [variant] and [theme] parameters.
  ///
  /// The [width] parameter defaults to 300. This is the width of the button.
  ///
  /// The [borderRadius] parameter defaults to 8.
  /// This is the border radius of the button.
  ///
  /// Returns a `SocialAuthBtn` configured for Facebook.
  factory SocialAuthBtn.facebook({
    required VoidCallback onPressed,
    FacebookTypeVariants variant = FacebookTypeVariants.normal,
    FacebookThemeVariants theme = FacebookThemeVariants.light,
    double width = 300,
    double borderRadius = 8,
    bool loading = false,
    Color? loaderColor,
  }) {
    final isDark = theme == FacebookThemeVariants.dark;
    var borderSide = BorderSide.none;
    var backgroundColor = isDark ? Colors.black : facebookBrandColor;
    var textColor = Colors.white;

    Color? iconColor;
    if (variant == FacebookTypeVariants.outlined) {
      borderSide = BorderSide(
        color: isDark ? Colors.black : facebookBrandColor,
      );
      backgroundColor = Colors.white;
      textColor = isDark ? Colors.black : facebookBrandColor;
      iconColor = isDark ? Colors.black : facebookBrandColor;
    }
    return SocialAuthBtn(
      text: 'Login in with Facebook',
      icon: 'assets/logos/facebook.png',
      iconColor: iconColor,
      onPressed: onPressed,
      width: width,
      borderRadius: borderRadius,
      backgroundColor: backgroundColor,
      textColor: textColor,
      borderSide: borderSide,
      package: 'social_auth_btn_kit',
      loading: loading,
      loaderColor: loaderColor,
    );
  }

  /// Creates a new instance of `SocialAuthBtn` for Google.
  ///
  /// The [onPressed] parameter must not be null.
  /// This is the callback that will be called when the button is pressed.
  ///
  /// The [theme] parameter defaults to `GoogleThemeVariants.light`.
  /// This determines the theme of the Google button.
  ///
  /// The button's border, background color, text color, and icon color
  /// are determined based on the [theme] parameter.
  ///
  /// The [width] parameter defaults to 300. This is the width of the button.
  ///
  /// The [borderRadius] parameter defaults to 8.
  /// This is the border radius of the button.
  ///
  /// Returns a `SocialAuthBtn` configured for Google.
  factory SocialAuthBtn.google({
    required VoidCallback onPressed,
    GoogleThemeVariants theme = GoogleThemeVariants.light,
    double width = 300,
    double borderRadius = 8,
    bool loading = false,
    Color? loaderColor,
  }) {
    var backgroundColor = googleBgColorLight;
    var textColor = googleLblColorLight;
    var borderSide = BorderSide.none;

    if (theme == GoogleThemeVariants.light) {
      borderSide = const BorderSide(
        color: googleBorderColorLight,
      );
    }

    if (theme == GoogleThemeVariants.dark) {
      borderSide = const BorderSide(
        color: googleBorderColorDark,
      );
      backgroundColor = googleBgColorDark;
      textColor = googleLblColorDark;
    }

    if (theme == GoogleThemeVariants.neutral) {
      backgroundColor = googleBgColorNeutral;
      textColor = googleLblColorNeutral;
    }

    return SocialAuthBtn(
      text: 'Sign in with Google',
      icon: 'assets/logos/google.png',
      onPressed: onPressed,
      width: width,
      borderRadius: borderRadius,
      borderSide: borderSide,
      backgroundColor: backgroundColor,
      textColor: textColor,
      package: 'social_auth_btn_kit',
      loading: loading,
      loaderColor: loaderColor,
    );
  }

  /// Creates a new instance of `SocialAuthBtn` for Apple.
  ///
  /// The [onPressed] parameter must not be null.
  /// This is the callback that will be called when the button is pressed.
  ///
  /// The [theme] parameter defaults to `AppleThemeVariants.white`.
  /// This determines the theme of the Apple button.
  ///
  /// The [variant] parameter defaults to `AppleTypeVariants.normal`.
  /// This determines the type of Apple button to display.
  ///
  ///
  factory SocialAuthBtn.apple({
    required VoidCallback onPressed,
    AppleThemeVariants theme = AppleThemeVariants.white,
    AppleTypeVariants variant = AppleTypeVariants.normal,
    double width = 300,
    double borderRadius = 8,
    bool loading = false,
    Color? loaderColor,
    Key? key,
  }) {
    var borderSide = BorderSide.none;

    var txtColor = appleLblColorLight;
    var bgColor = Colors.white;
    Color? iconColor;

    if (theme == AppleThemeVariants.black) {
      txtColor = Colors.white;
      bgColor = Colors.black;
      iconColor = Colors.white;
    }

    if (variant == AppleTypeVariants.outlined) {
      borderSide = const BorderSide();
    }

    return SocialAuthBtn(
      text: 'Sign in with Apple',
      icon: 'assets/logos/apple.png',
      iconColor: iconColor,
      onPressed: onPressed,
      width: width,
      borderRadius: borderRadius,
      borderSide: borderSide,
      key: key,
      textColor: txtColor,
      backgroundColor: bgColor,
      loading: loading,
      loaderColor: loaderColor,
      package: 'social_auth_btn_kit',
    );
  }

  final String text;
  final String icon;
  final String? package;
  final Color? iconColor;
  final Color backgroundColor;
  final Color textColor;
  final double width;
  final double borderRadius;
  final BorderSide borderSide;
  final VoidCallback onPressed;
  final bool loading;
  final Color? loaderColor;

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: width,
      child: ElevatedButton(
        onPressed: onPressed,
        style: ElevatedButton.styleFrom(
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(borderRadius),
            side: borderSide,
          ),
          elevation: 0,
          backgroundColor: backgroundColor,
        ),
        child: Padding(
          padding: const EdgeInsets.all(12),
          child: loading
              ? Center(
                  child: SizedBox(
                    width: 18,
                    height: 18,
                    child: CircularProgressIndicator(
                      backgroundColor: loaderColor,
                    ),
                  ),
                )
              : Row(
                  children: [
                    Image.asset(
                      icon,
                      package: package,
                      width: 24,
                      color: iconColor,
                    ),
                    const SizedBox(width: 24),
                    Expanded(
                      child: Text(
                        text,
                        style: TextStyle(
                          fontSize: 14,
                          fontWeight: FontWeight.w500,
                          color: textColor,
                        ),
                      ),
                    ),
                  ],
                ),
        ),
      ),
    );
  }
}


lib/social_auth_btn_variants.dart
/// FacebookTypeVariants
/// This enum determines the type of Facebook button to display.
/// The [normal] variant displays a button with a Facebook logo and text.
/// The [outlined] variant displays a button with a outline with brand color.
/// The [normal] variant is the default variant.
enum FacebookTypeVariants {
  normal,
  outlined,
}

/// FacebookThemeVariants
/// This enum determines the theme of the Facebook button.
/// The [light] variant displays a button with a light theme.
/// The [dark] variant displays a button with a dark theme.
/// The [light] variant is the default variant.
enum FacebookThemeVariants {
  light,
  dark,
}

/// GoogleTypeVariants
/// This enum determines the type of Google button to display.
/// The [light] variant displays a button with a Google logo and text.
/// The [dark] variant displays a button in dark theme.
/// The [neutral] variant displays a grey button
enum GoogleThemeVariants {
  light,
  dark,
  neutral,
}

/// AppleTypeVariants
/// This enum determines the type of Apple button to display.
/// The [normal] variant displays a button with a Apple logo and text.
/// The [outlined] variant displays a button with a outline with brand color.
enum AppleTypeVariants {
  normal,
  outlined,
}

/// AppleThemeVariants
/// This enum determines the theme of the Apple button.
/// The [white] variant displays a button with a white theme.
/// The [black] variant displays a button with a black theme.
/// The [white] variant is the default variant.
enum AppleThemeVariants {
  black,
  white,
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:social_auth_btn_kit/social_auth_btn_kit.dart';
import 'package:social_auth_btn_kit/social_auth_btn_variants.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Social Auth Btn Kit Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Social Auth Btn Kit Demo'),
        ),
        backgroundColor: Colors.white,
        body: Padding(
          padding: const EdgeInsets.all(16.0),
          child: ListView(
            children: [
              Center(
                child: SocialAuthBtn.facebook(
                  variant: FacebookTypeVariants.outlined,
                  onPressed: () {
                    debugPrint("DEBUG: Facebook Btn Pressed");
                  },
                  loading: true,
                ),
              ),
              const SizedBox(height: 16),
              Center(
                child: SocialAuthBtn.facebook(
                  onPressed: () {
                    debugPrint("DEBUG: Facebook Btn Pressed");
                  },
                ),
              ),
              const SizedBox(height: 16),
              Center(
                child: SocialAuthBtn.facebook(
                  theme: FacebookThemeVariants.dark,
                  onPressed: () {
                    debugPrint("DEBUG: Facebook Btn Pressed");
                  },
                ),
              ),
              const SizedBox(height: 16),
              Center(
                child: SocialAuthBtn.facebook(
                  variant: FacebookTypeVariants.outlined,
                  theme: FacebookThemeVariants.dark,
                  onPressed: () {
                    debugPrint("DEBUG: Facebook Btn Pressed");
                  },
                ),
              ),
              const SizedBox(height: 16),
              Center(
                child: SocialAuthBtn.google(
                  onPressed: () {
                    debugPrint("DEBUG: Google Btn Pressed");
                  },
                ),
              ),
              const SizedBox(height: 16),
              Center(
                child: SocialAuthBtn.google(
                  onPressed: () {
                    debugPrint("DEBUG: Google Btn Pressed");
                  },
                  theme: GoogleThemeVariants.dark,
                ),
              ),
              const SizedBox(height: 16),
              Center(
                child: SocialAuthBtn.google(
                  onPressed: () {
                    debugPrint("DEBUG: Google Btn Pressed");
                  },
                  theme: GoogleThemeVariants.neutral,
                ),
              ),
              const SizedBox(height: 16),
              Center(
                child: SocialAuthBtn.apple(
                  onPressed: () {
                    debugPrint("DEBUG: Apple Btn Pressed");
                  },
                ),
              ),
              const SizedBox(height: 16),
              Center(
                child: SocialAuthBtn.apple(
                  onPressed: () {
                    debugPrint("DEBUG: Apple Btn Pressed");
                  },
                  variant: AppleTypeVariants.outlined,
                ),
              ),
              const SizedBox(height: 16),
              Center(
                child: SocialAuthBtn.apple(
                  onPressed: () {
                    debugPrint("DEBUG: Apple Btn Pressed");
                  },
                  theme: AppleThemeVariants.black,
                ),
              ),
              const SizedBox(height: 16),
              Center(
                child: SocialAuthBtn(
                  onPressed: () {
                    debugPrint("DEBUG: Instagram Btn Pressed");
                  },
                  icon: 'assets/instagram.png',
                  text: 'Connect Instagram',
                  borderSide: const BorderSide(),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
